huntsman/observatory.py

```diff
@@ -512,3 +378,157 @@ def _create_flat_field_observation(self, alt=None, az=None,
+        exptimes = {cam_name: [1. * u.second] for cam_name in cameras.keys()}
+
+        # Loop until conditions are met to finish flat-fielding
+        exposure_count = 0
+        while True:
+
+            # Update the exposure count
+            exposure_count += 1
+
+            # Get the FITS headers with a common start time
+            start_time = utils.current_time()
+            fits_headers = self.get_standard_headers(observation=observation)
+            fits_headers['start_time'] = utils.flatten_time(start_time)
+
+            # Take the flat field observations (blocking)
+            next_exptimes = {c: exptimes[c][-1] for c in cameras.keys()}
+            camera_events = self._take_flat_fields(next_exptimes, observation,
+                                                   fits_headers=fits_headers)
+
+            # Check the counts for each image
+            is_saturated = False
+            for cam_name, info in camera_events.items():
+                filename = info['filename']
+                self.logger.debug("Checking counts for {}".format(filename))
+
+                # Load the image we just took
+                try:
+                    data = fits.getdata(filename)
+                except FileNotFoundError:
+                    data = fits.getdata(filename + '.fz')
+
+                # Calculate average counts per pixel
+                mean_counts, _, _ = stats.sigma_clipped_stats(data)
+                mean_counts = mean_counts - bias
+                self.logger.debug("Counts: {}".format(mean_counts))
+                if (mean_counts < min_counts) or (mean_counts > max_counts):
+                    self.logger.debug("Counts outside min/max range, should be discarded.")
+
+                # Check if camera is saturated
+                if mean_counts >= max_counts:
+                    is_saturated = True
+
+                # Get suggested exposure time
+                elapsed_time = (utils.current_time() - start_time).sec
+                exptime = exptimes[cam_name][-1].value * (target_adu / mean_counts)
+                # Account for the sky changing brightness
+                sky_factor = 2.0 ** (elapsed_time / 180.0)
+                if self.past_midnight:
+                    exptime = exptime / sky_factor
+                else:
+                    exptime = exptime * sky_factor
+                exptime = int(max(min_exptime, exptime))
+                exptimes[cam_name].append(exptime * u.second)
+                self.logger.debug("Elapsed time: {}".format(elapsed_time))
+                self.logger.debug("Suggested exptime for {}: {}".format(cam_name, exptime))
+
+            # Stop flats if any time is greater than max
+            exptime_too_long = [t[-1].value >= max_exptime for t in exptimes.values()]
+            if any(exptime_too_long) and not self.past_midnight:
+                self.logger.debug("Exposure times greater than max, stopping flat fields.")
+                break
```

Comment:
Don't we still want to check for this in the morning?  Or is it just that we don't know what "too long" looks like right now. 

Or I guess maybe it is handled by the saturation anyway?